### DISCUSSION TITLE
Gas tradeoffs in the CEI model.

### DISCUSSION THREAD

**Discussion by ebok21** [Cyfrin/foundry-full-course-cu#869] [General]
Why do we perform the transfer operation after the updating of the mapping?
If the transaction were to fail, the mapping would hold wrong info or does the revert undo the changes made?
And even if the changes are undone, is the gas spent recovered? Why not just perform the interactions before the effects? @PatrickAlphaC 

```solidity
// Follows CEI
// Checks in function definition using modifiers
function depositCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        external
        moreThanZero(amountCollateral)
        isAllowedToken(tokenCollateralAddress)
        nonReentrant
    {
       // Effects - Update the mapping and emit event
        s_collateralDeposited[msg.sender][tokenCollateralAddress] += amountCollateral;
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);

       // Perform the token transfer
        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
```

**Comment by PatrickAlphaC** [comment]
We want to follow the [CEI/FREI-PI](https://www.nascent.xyz/idea/youre-writing-require-statements-wrong) pattern. Whenever we make a call to a contract outside our protocol, we want it to be the last step, otherwise, we are susceptible to the dreaded reentrancy attack!

**Reply by ebok21** [reply]
thanks a ton!

**Comment by PatrickAlphaC** [comment]
Check out [this attack](https://solidity-by-example.org/hacks/re-entrancy/)

**Reply by ebok21** [reply]
yep will do
